### PR TITLE
[#10938] Support copying courses

### PR DESCRIPTION
**Pull request by ypinhsuan**
Part of #10938, Closes #11012

**Outline of Solution**
Original credits to @LeeBadal, @heeenkie, @Grunler, @MatildaAR, @Olle251 for implementation and frontend tests

A copy course option similar to copy feedback sessions
Added an option to allow instructors to copy from other courses when adding new course

There should be a dropdown list for `Course to copy from` but somehow my screen recording is not capturing it
gif
TEAMMATES-Online-Peer-Feedback-E

### CODE REVIEW

File: src/web/app/components/copy-course-modal/copy-course-modal.component.ts
Comment (daongochieu2810) [inline src/web/app/components/copy-course-modal/copy-course-modal.component.ts:108]: you can use `this.chosenFeedbackSessions.clear()` here

File: src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts
Comment (daongochieu2810) [inline src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts:115]: I think we should declare this method as `async` so it will not block for long in case of coying multiple big courses

File: src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts
Comment (daongochieu2810) [inline src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts:134]: use `await` after declaring this method as `async`

**Review by daongochieu2810 — COMMENTED** [review]
Very good overall! Left some comments

File: src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts
Comment (moziliar) [reply src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts:115]: This at the same time means we are also allowing concurrent connection to the backend. How's the resource consumption of this change?

File: src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts
Comment (moziliar) [inline src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:352]: Discussion: is there a way to replace this with something like `feedbackSession.toCreationReqWithName(newSessionName)`?

File: src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts
Comment (moziliar) [inline src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:387]: Likewise, any possibility to abstract away this?
Otherwise, it is fine to leave as is; just somewhat unnecessarily chunky here reading as "copying everything from `feedbackQuestion` to a new struct that inherits all the field values => semantically equivalent to a `clone()`

**Review by moziliar — CHANGES_REQUESTED** [review]
Thanks for the original authors' efforts and @ypinhsuan's effort in curating it towards completion.

Really nicely written code from what I read. Would've been even nicer to have a gif/video to showcase the entire workflow in this PR too.

Have some comments below and some points for discussion.

File: src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts
Comment (daongochieu2810) [reply src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts:115]: yes I agree with disabling the copy course button while copying. Or we can turn it into a cancel button

Comment (moziliar) [reply src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts:115]: This sounds like a UX question😅I'm no expert in it, but just raising the resource issue.
Can do either way! Cancel button might involve rollback btw.

Comment (ypinhsuan) [reply src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts:115]: Is this ok? The copy button is disabled when copying.
ezgif com-gif-maker

File: src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts
Comment (ypinhsuan) [reply src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:271]: I don't think there will be such a case where `courseId` is a falsy value, so I will remove this part

File: src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts
Comment (ypinhsuan) [reply src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:387]: ok! will split this into a separate function too

**Comment by ypinhsuan** [comment]
Added gif in PR description and some frontend tests

File: src/web/app/components/copy-course-modal/copy-course-modal.component.ts
Comment (moziliar) [inline src/web/app/components/copy-course-modal/copy-course-modal.component.ts:113]: Am wondering if this can be outside the class? It is also fine being inside too:p

File: src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts
Comment (moziliar) [reply src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts:116]: Sorry for not getting back on this.
@damithc's take on loading UI would be to show the user some progress when it takes sometime to load.
Is there a way to perform this? I remember we are copying session by session (in the future we might need to improve this, but so far no such need as this is not a frequent use case).

File: src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts
Comment (moziliar) [inline src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts:135]: Hmm can I clarify which function call inside is async? `showErrorToast`?

File: src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts
Comment (moziliar) [inline src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts]: Sorry regarding this: do you think there should still be a defensive mechanism in place?
Was commenting on this as the interpolated `courseId` is most likely redundant, but I wasn't rejecting the possibility of such error happening.
Can think of other places that might reuse this function.

File: src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts
Comment (moziliar) [inline src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:394]: `toFbQuestionCreationReq` or `newFbQuestionCreationReq`? Making it an action sounding name is probably better.

**Review by moziliar — DISMISSED** [review]
Added some more comments. Can discuss against my comments if they don't make sense:)
I am largely ok with the current logic. Might need to do some user testing as well before finalising it.

Do resolve the review threads too.

On another note: it is probably better to have more meaningful comment names than `Apply suggested changes`; for the reviewer, I can review by commit to see the exact changes under the commit name; for a developer, I can also `reset` or `revert` certain changes easily with a clearer defined bound.
I know it is hard to keep this up all the time, and that this PR is likely gonna be squashed:p but it's a good habit to have, and that the PR history is still around after squashed merge and can let future maintainer to see exactly what happened.

File: src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts
Comment (ypinhsuan) [reply src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts:135]: I am not very sure but my understanding of `await` is that it is used on `Promise` objects and in this case `modalRef.result` is a Promise. 
```
        try {
          await modalRef.result
          const result: CopyCourseModalResult = await modalRef.result;
          this.copyCourseEvent.emit(result)
        } catch (resp: any) {
          this.statusMessageService.showErrorToast(resp.error.message);
        }
```
Would it be correct to change this part to the above?

File: src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts
Comment (ypinhsuan) [reply src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts]: I see. I agree it will be better to have defensive mechanism

File: src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts
Comment (ypinhsuan) [reply src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts:116]: Added a progress bar to show users the progress
New-video

Comment (moziliar) [reply src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts:116]: This is awesome! I love it!

@damithc prof, are you satisfied with this?
It might not stay should this feature becomes too popular and we subsequently make the endpoint granular (copy course and session all in one API call). But before that, I'd want to see this myself actually.

File: src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts
Comment (moziliar) [reply src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts]: As in: is there any reason to make this a blocking call?
`subscribe` itself is already a Promise-ish thing. I don't see why you need an await here even tho `modalRef.result` is a Promise.

File: src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts
Comment (moziliar) [inline src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts:133]: Looks like it will fan out into many API call at once. I'd keep an eye on it.
What if there are too many sessions to copy? Will it cause new instances to be spun up?
@rrtheonlyone

File: src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts
Comment (moziliar) [inline src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:312]: Btw, when you are copying feedback sessions, is it also considered `CopyingCourse`?
My doubt would be is it possible for the same create course to be fired the second time while the first is still creating sessions, which would result in 2x or 3x (depends on how many times the course is copied) and an unjustifiable backend load?

**Review by moziliar — COMMENTED** [review]
Left some comments.

Generally it feels good to me!

**Review by ccyccyccy — COMMENTED** [review]
Found a bug. Copying to a destination course id with the same course id as a currently existing course will lead to the page showing a success message even though it failed.
ezgif-4-59d45c04fe03

Also, the select all checkbox merges into the rest of it's selection. Can we separate them more? Something like this would be nice. 
image

File: src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts
Comment (ypinhsuan) [reply src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts]: Ok I see, then I think the async and await is not needed here.

File: src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts
Comment (ypinhsuan) [reply src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:312]: Copying feedback sessions is a part of `CopyingCourse`. When feedback sessions are still being copied, the copy button will be disabled so instructors cannot copy another course at the same time. But they can still add a new course with just course name and course id.

File: src/web/app/components/copy-course-modal/copy-course-modal.component.scss
Comment (ypinhsuan) [reply src/web/app/components/copy-course-modal/copy-course-modal.component.scss:1]: Not used, I will remove this

File: src/web/app/components/copy-course-modal/copy-course-modal.component.ts
Comment (ypinhsuan) [reply src/web/app/components/copy-course-modal/copy-course-modal.component.ts:48]: I will keep the type since lint is checking for typedef and I will fail the check if I remove them

**Comment by ypinhsuan** [comment]
Thank you for all the comments! I have updated the UI accordingly and fixed the bugs.
updates

**Comment by ccyccyccy** [comment]
> updates

Can we simply disable the "Copy" button with a warning text beside the input box to avoid this situation all together?

**Comment by ypinhsuan** [comment]
The input box will turn red if the course ID is already been used (includes archived and soft-deleted courses) and a toast message will be displayed. Users can only copy from active courses for now.
disable copy button

File: src/web/app/components/copy-course-modal/copy-course-modal.component.ts
Comment (ccyccyccy) [inline src/web/app/components/copy-course-modal/copy-course-modal.component.ts:74]: Can change variable name to `newCourseIdIsConflicting`. Currently is a little confusing.

Comment (ccyccyccy) [reply src/web/app/components/copy-course-modal/copy-course-modal.component.ts:74]: Also, this variable should be set back to false on focus on the input element.

File: src/web/app/components/copy-course-modal/copy-course-modal.component.scss
Comment (ccyccyccy) [inline src/web/app/components/copy-course-modal/copy-course-modal.component.scss:2]: Can override the margin here? The default margin is way too much. Something like 0.3rem?

File: src/web/app/components/copy-course-modal/copy-course-modal.component.ts
Comment (ccyccyccy) [inline src/web/app/components/copy-course-modal/copy-course-modal.component.ts:52]: When I mentioned using `.map`, i meant something like `this.timezones = Object.entries(this.timezoneService.getTzOffsets()).map(...)`, so that you don't have to use `.push()`

**Review by ccyccyccy — COMMENTED** [review]
Just some minor issues for me. Otherwise, all looks good.

File: src/web/app/components/copy-course-modal/copy-course-modal.component.scss
Comment (ypinhsuan) [reply src/web/app/components/copy-course-modal/copy-course-modal.component.scss:2]: I have reduced the margin for the border to look like this:
ezgif com-gif-maker

**Review by daongochieu2810 — APPROVED** [review]

**Comment by rrtheonlyone** [comment]
For your kind review: either @ccyccyccy or @moziliar, when you have time. Let me have a final review before merging.

**Review by rrtheonlyone — CHANGES_REQUESTED** [review]
Nice one 💯 I like the UI here and the frontend looks good.

There are two aspects that I am not a fan of though:
1. We are being too fine-grained in the copying. We should not need to make this many round-trips. 
2. We are duplicating the feedback session copying logic in 2 places. This is prone to errors down the road.

**Important**: We also need to update E2E tests in `InstructorCoursesPageE2ETest` for this feature before we can merge in.

This is my proposal for this. Let me know what you think @ypinhsuan 

## Optimized Copying

### Method

1. We should augment the `CreateFeedbackSessionAction` endpoint to have an additional field in the body e.g. like 'toCopySessionId` or something
2. Upon seeing this field the backend can then do this:
   - Get the old session and create a new session
   - Bulk get all questions corresponding to old session and bulk create new questions in new session
3. Return `FeedbackSessionData` 
4. We can then update `feedbackSessionsService` and use the new method in both places (no more duplication)

### Rationale

Step (2) should not be an expensive operation since there are only 4 database operations. Even if there are 1000 questions, I suspect it should easily finish within a reasonable amount of time (< 1 minute). However, this is only a guess and I do not have tests to prove it. It is worth a try though.

**Comment by ypinhsuan** [comment]
I agree that there seem to be too many round-trips when copying a session. The optimized copying looks good and I will try implementing it.

**Comment by ypinhsuan** [comment]
I have changed the way feedback sessions are copied and added e2e test for copying course.

File: src/main/java/teammates/ui/request/FeedbackSessionCreateRequest.java
Comment (rrtheonlyone) [inline src/main/java/teammates/ui/request/FeedbackSessionCreateRequest.java:40]: We need to update the validation - I think if `toCopyCourseId` is not null then `oldSessionName` should not be null as well?

Can we also rename `oldSessionName` to `toCopySessionName`?

File: src/main/java/teammates/ui/webapi/CreateFeedbackSessionAction.java
Comment (rrtheonlyone) [inline src/main/java/teammates/ui/webapi/CreateFeedbackSessionAction.java:104]: let's think about this for a second - why do we need to validate again if the old questions have already been validated? Tell me a scenario in which we run into a validation error here

**Review by rrtheonlyone — CHANGES_REQUESTED** [review]
Excellent Work! Give me a ping (on slack) once you fix the request changes and I will prioritise the review immediately.

Minor comments below plus:
- Can we update `CreateFeedbackSessionActionTest` for the new behaviour we just introduced?
- Fix merge conflicts in the E2E tests

File: src/main/java/teammates/ui/request/FeedbackSessionCreateRequest.java
Comment (ypinhsuan) [reply src/main/java/teammates/ui/request/FeedbackSessionCreateRequest.java:40]: I have updated the validation and renamed to `toCopySessionName`

File: src/main/java/teammates/ui/webapi/CreateFeedbackSessionAction.java
Comment (ypinhsuan) [reply src/main/java/teammates/ui/webapi/CreateFeedbackSessionAction.java]: I removed the part that checks for the validity of giver and recipient.
If a feedback session is copied to a course with different number of students / teams, the `numOfGeneratedMsqChoices` would be different. So the part that validates question details is needed to update `numOfGeneratedMsqChoices` .

Comment (wkurniawan07) [reply src/main/java/teammates/ui/webapi/CreateFeedbackSessionAction.java]: @damithc I need to raise this here. This is a problem I've been aware of for quite some time. (hello #8453)

This part here validates that the max or min number of choices (whichever ones configured) is/are within the range of the number of generated choices (student, team, instructor, etc.). At the first glance, it seems sensible. However, this creates an unnecessary coupling between the question and the course structure.

Consider this:
- A question with min. 5 choice, generated from instructors, is created, in a course with 6 instructors
- 2 instructors are removed from that course
- This question is now unanswerable, as the validation is done only during question creation/update, but not when the coupled course structure is changed

And in this very example (copying course), any MSQ question with minimum number of selected choice generated from students/teams is not copiable, because a new course doesn't have any student list to be generated from!

I suggest to just remove this validation.

Comment (damithc) [reply src/main/java/teammates/ui/webapi/CreateFeedbackSessionAction.java]: > I suggest to just remove this validation. Not because this validation creates the problem or removing this validation solves the problem, but because this validation doesn't solve the problem it intends to.

I agree. Let's do that.

File: src/web/app/components/copy-course-modal/copy-course-modal.component.html
Comment (wkurniawan07) [inline src/web/app/components/copy-course-modal/copy-course-modal.component.html:1]: Use 2 spaces for HTML file

File: src/web/app/components/copy-course-modal/copy-course-modal.component.html
Comment (wkurniawan07) [inline src/web/app/components/copy-course-modal/copy-course-modal.component.html:26]: `64` is a magic number

File: src/main/java/teammates/ui/request/FeedbackSessionCreateRequest.java
Comment (wkurniawan07) [inline src/main/java/teammates/ui/request/FeedbackSessionCreateRequest.java:45]: - The validation message doesn't match
- "To copy session name" is not a noun (it's a verb). Should be "Session name to be copied from".

File: src/main/java/teammates/ui/webapi/CreateFeedbackSessionAction.java
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/CreateFeedbackSessionAction.java:114]: A valid question in one session should be valid in any other session, so if this happens, something is not right and it's not due to user. Need a `log.severe` here instead of throwing this exception.

File: src/web/app/components/copy-course-modal/copy-course-modal.component.ts
Comment (wkurniawan07) [inline src/web/app/components/copy-course-modal/copy-course-modal.component.ts:75]: This message makes it sound like you're comparing the name with all courses in the DB, not just all courses from the user. Is it really what's being done here?

File: src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:363]: `courseModel` needs not be a variable? Can just return the previous line directly.

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
Not really looking into detail, but I do have some questions/concerns.

File: src/web/app/components/copy-course-modal/copy-course-modal.component.ts
Comment (ypinhsuan) [reply src/web/app/components/copy-course-modal/copy-course-modal.component.ts:75]: This part checks only for courses that belong to the instructor, I will update the message.

File: src/main/java/teammates/ui/request/FeedbackSessionCreateRequest.java
Comment (ypinhsuan) [reply src/main/java/teammates/ui/request/FeedbackSessionCreateRequest.java]: Updated the message

**Review by rrtheonlyone — APPROVED** [review]
LGTM 💯 Thanks for the work put in here!

**Review by wkurniawan07 — APPROVED** [review]

**Pull request merged**
